packages/core/src/client.ts

```diff
@@ -1969,6 +1969,31 @@ export class MedplumClient extends EventTarget {
       this.createResource(resource, options)) as Promise<T>;
   }
 
+  /**
+   * Upsert a resource: update it in place if it exists, otherwise create it.  This is done in a single, transactional
+   * request to guarantee data consistency.
+   * @param resource - The resource to update or create.
+   * @param query - A FHIR search query to uniquely identify the resource if it already exists.
+   * @param options  - Optional fetch options.
+   * @returns The updated/created resource.
+   */
+  async upsertResource<T extends Resource>(resource: T, query: string, options?: MedplumRequestOptions): Promise<T> {
```

Comment:
@mattwiller , for the `query` param, can we support the `query: QueryTypes`, similar to what we use in `searchResources`. 

Being able to pass a dictionary in can be very handy

Reply:
This is a good call.  See `searchResources`.  You can then use `fhirSearchUrl` to generate the URL as well.

```ts
    const url = this.fhirSearchUrl(resourceType, query);
```